### PR TITLE
fix: Add validation to prevent numeric values in patient names

### LINKED ISSUE #2602: Disallow numeric values in patient names
Add backend validations too for issue mentioned in: https://github.com/ohcnetwork/care_fe/issues/9143

### PR DESCRIPTION
**Pull request by areebahmeddd**
## Proposed Changes

- Added backend validation to disallow numeric values in patient names.
- Added test cases for valid and invalid patient names.

### Associated Issue

- Fixes #2602: Adds validation to ensure patient names do not contain numeric values, preventing invalid patient creation.

## Merge Checklist

- [x] Tests added/fixed
- [x] Update docs in `/docs`
- [x] Linting Complete
- [ ] Any other necessary step

@ohcnetwork/care-backend-maintainers @ohcnetwork/care-backend-admins

## Summary by CodeRabbit

- **New Features**
	- Enhanced validation for patient names to ensure they consist solely of alphabetic characters and spaces.
	- Improved checks to require either year of birth or date of birth during patient record creation and updates.
	- Updated patient names in the dataset for clarity and organization.
	- Updated user details for improved identification and compliance.

- **Bug Fixes**
	- Clearer error messages for validation failures regarding vaccination details.

- **Tests**
	- Added tests to verify successful registration for valid patient names and appropriate error responses for invalid names.
  
- **Chores**
	- Introduced automated code quality checks using Qodana in the CI/CD pipeline.
	- Added configuration for Qodana to ensure compliance with dependency licenses.

### CODE REVIEW

File: care/facility/api/serializers/patient.py [hunk 1]
```diff
@@ -210,6 +210,15 @@ def validate_year_of_birth(self, value):
             raise serializers.ValidationError(msg)
         return value
 
+    def validate_name(self, value):
+        if value is None:
+            raise serializers.ValidationError("Patient name is required.")
+        if not value.strip():
+            raise serializers.ValidationError("Patient name cannot be empty.")
```
Comment (shivankacker) [inline care/facility/api/serializers/patient.py]: I think the model will handle this by default, no need to add these validations in the serializer.

File: care/facility/api/serializers/patient.py [hunk 2]
```diff
@@ -210,6 +210,15 @@ def validate_year_of_birth(self, value):
...
+        if any(char.isdigit() for char in value):
```
Comment (shivankacker) [inline care/facility/api/serializers/patient.py]: This will prevent numbers, but what about dots, commas, underscores? I think we should use regex here.

**Review by shivankacker — CHANGES_REQUESTED** [review]

File: care/facility/api/serializers/patient.py (hunk 2 above)
Comment (areebahmeddd) [reply care/facility/api/serializers/patient.py]: My thoughts exactly but then i decided to stick with what the issue title was haha

File: care/facility/api/serializers/patient.py [hunk 3]
```diff
@@ -240,6 +246,9 @@ def check_external_entry(self, srf_id):
             )
 
     def create(self, validated_data):
+        name = validated_data.get("name")
+        if name:
+            self.validate_name(name)
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion
```
this is not required as create and update already get the validated data

File: care/facility/api/serializers/patient.py [hunk 4]
```diff
@@ -300,6 +309,9 @@ def create(self, validated_data):
         return patient
 
     def update(self, instance, validated_data):
+        name = validated_data.get("name")
+        if name:
+            self.validate_name(name)
```
Comment (sainak) [inline care/facility/api/serializers/patient.py]:
```suggestion
```
this is not required as create and update already get the validated data

**Review by sainak — CHANGES_REQUESTED** [review]

File: care/facility/api/serializers/patient.py (hunk 3 above)
Comment (areebahmeddd) [reply care/facility/api/serializers/patient.py]: I see, understood :)
Could you confirm the latest commit? @sainak

**Comment by rithviknishad** [comment]
Closing as per https://github.com/ohcnetwork/care_fe/issues/9143#issuecomment-2491409154

**Pull request closed without merging**